tests/federation_room_join_partial_state_test.go

```diff
@@ -3295,6 +3295,85 @@ func TestPartialStateJoin(t *testing.T) {
 		}
 	})
 
+	t.Run("User directory is correctly updated once state re-sync completes", func(t *testing.T) {
+		rocky := deployment.RegisterUser(t, "hs1", "rocky", "pass", false)
+
+		server := createTestServer(t, deployment)
+		cancel := server.Listen()
+		defer cancel()
+		serverRoom := createTestRoom(t, server, rocky.GetDefaultRoomVersion(t))
+
+		// start a partial state join
+		psjResult := beginPartialStateJoin(t, server, serverRoom, rocky)
+		defer psjResult.Destroy(t)
+
+		// sanity check - before the state has completed syncing state we would expect rocky to show up
+		reqBody := client.WithJSONBody(t, map[string]interface{}{
+			"search_term": rocky.UserID,
+		})
+		res := rocky.MustDoFunc(t, "POST", []string{"_matrix", "client", "v3", "user_directory", "search"}, reqBody)
+		must.MatchResponse(t, res, match.HTTPResponse{
+			StatusCode: 200,
+			JSON: []match.JSON{
+				match.JSONKeyEqual("results.0.user_id", "@rocky:hs1"),
+			}})
+
+		// ...but not charlie's
+		reqBody2 := client.WithJSONBody(t, map[string]interface{}{
+			"search_term": "charlie",
+		})
+		res2 := rocky.MustDoFunc(t, "POST", []string{"_matrix", "client", "v3", "user_directory", "search"}, reqBody2)
+		must.MatchResponse(t, res2, match.HTTPResponse{
+			StatusCode: 200,
+			JSON: []match.JSON{
+				match.JSONKeyEqual("results", [0]string{}),
+			}})
+
+		// finish syncing the state
+		psjResult.FinishStateRequest()
+		awaitPartialStateJoinCompletion(t, psjResult.ServerRoom, rocky)
+
+		// the rooms stats are updated by a background job in Synapse which is not guaranteed to have completed by the time
+		// the state sync has completed. We check for up to 3 seconds that the job has completed, after which Charlie
+		// the job should have finished and Charlie and Derek should be visible in the user directory
+		time.Sleep(time.Second * 3)
```

Comment:
In the interest of expediency I stuck a sleep here. I attempted on lines 3363-3374 (I commented out the attempt in case it's helpful to see) to use `client.WithRetryUntil` to loop until the condition was met, but I could not figure out how to make that function work with a request body, which is necessary for this API (I think?). It worked for the first iteration but then complained about an empty body for the next iteration, my best guess is that the request body is being consumed on the first attempt? I tried a few different formulations to feed the request body in but failed, any insight here would be appreciated.

Reply:
> the request body is being consumed on the first attempt

That looks like what is happening. `req.Body` is set to an `io.Reader` containing the JSON. At the end of the first request, the read position lies at the end of the JSON, and so the retry reads 0 bytes.

There's a `GetBody` field that can be set on `http.Request`. If we modify `WithRawBody`, we can get the retries to work:
```diff
diff --git a/internal/client/client.go b/internal/client/client.go
index 7ffacff..05dc056 100644
--- a/internal/client/client.go
+++ b/internal/client/client.go
@@ -522,6 +522,10 @@ func (c *CSAPI) GetDefaultRoomVersion(t *testing.T) gomatrixserverlib.RoomVersio
 // WithRawBody sets the HTTP request body to `body`
 func WithRawBody(body []byte) RequestOpt {
        return func(req *http.Request) {
-               req.Body = ioutil.NopCloser(bytes.NewBuffer(body))
+               req.Body = ioutil.NopCloser(bytes.NewReader(body))
+               req.GetBody = func() (io.ReadCloser, error) {
+                       r := bytes.NewReader(body)
+                       return ioutil.NopCloser(r), nil
+               }
                // we need to manually set this because we don't set the body
                // in http.NewRequest due to using functional options, and only in NewRequest
                // does the stdlib set this for us.
```

We replace `NewBuffer` with `NewReader`, because the docs say
> NewBuffer creates and initializes a new Buffer using buf as its initial contents. The new Buffer takes ownership of buf, and the caller should not use buf after this call.

Reply:
I put this change in this PR, let me know if it should be in a separate PR instead.